### PR TITLE
[FLINK-25035][runtime] Shuffle service supports consuming subpartition range

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change
Let shuffle service supports consuming subpartition range. In adaptive batch scheduler, the shuffle service needs to support a SingleInputGate to consume  a certain range of subpartitions.

## Brief change log
68e3afaf272233bd0ca180eba092d06325fd2533 Move consumedSubpartitionIndex from SingleInputGate to InputChannel
41eeab31ee52a29d71acf018f9d1ef5a7750bb85 SingleInputGate supports consuming subpartition range

## Verifying this change
This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

**Review by zhuzhurk — COMMENTED** [review]
Thanks for opening this PR! @wanglijie95 
The changes of commit 12880feeedbbba9d3beb93225843218f4d6776f6 looks generally good to me. I only have 2 minor comments and please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java [hunk 1]
```diff
@@ -54,6 +54,8 @@
 
     protected final ResultPartitionID partitionId;
 
+    protected final int subpartitionIdx;
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java]: fixed

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java [hunk 2]
```diff
@@ -173,20 +177,35 @@ private void createInputChannels(
             String owningTaskName,
             InputGateDeploymentDescriptor inputGateDeploymentDescriptor,
             SingleInputGate inputGate,
+            InputGateDeploymentDescriptor.SubpartitionIndexRange subpartitionRange,
             InputChannelMetrics metrics) {
         ShuffleDescriptor[] shuffleDescriptors =
                 inputGateDeploymentDescriptor.getShuffleDescriptors();
 
         // Create the input channels. There is one input channel for each consumed partition.
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java]: fixed

**Comment by wanglijie95** [comment]
> Thanks for opening this PR! @wanglijie95 The changes of commit [12880fe](https://github.com/apache/flink/commit/12880feeedbbba9d3beb93225843218f4d6776f6) looks generally good to me. I only have 2 minor comments and please take a look.

Thanks for review @zhuzhurk .  I've addressed all your comments and split the changes into 2 commits, please review again.

**Comment by wanglijie95** [comment]
@wsry This change involves some modifications to the network. Would you review it?

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 3]
```diff
@@ -645,13 +616,13 @@ public void testRequestBackoffConfiguration() throws Exception {
                                     throw new RuntimeException(e);
                                 }
                             });
-            InputChannel localChannel = channelMap.get(partitionIds[0]);
+            InputChannel localChannel = channelMap.get(new SubpartitionInfo(partitionIds[0], 0));
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:619]: It's better to create a testing method `createSubpartitionInfo(partitionId)`.

**Comment by wanglijie95** [comment]
@flinkbot run azure

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java (hunk 3 above)
Comment (wanglijie95) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java]: fixed

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java [hunk 4]
```diff
@@ -94,6 +97,7 @@ protected InputChannel(
         this.inputGate = checkNotNull(inputGate);
         this.channelInfo = new InputChannelInfo(inputGate.getGateIndex(), channelIndex);
         this.partitionId = checkNotNull(partitionId);
+        this.consumedSubpartitionIndex = consumedSubpartitionIndex;
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:100]: nit: maybe add an ```>=0``` argument check

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java [hunk 5]
```diff
@@ -168,8 +176,7 @@ protected void notifyBufferAvailable(int numAvailableBuffers) throws IOException
      * <p>The queue index to request depends on which sub task the channel belongs to and is
      * specified by the consumer of this channel.
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:177]: ```and is specified by the consumer of this channel.``` This java doc seems out of date. Maybe delete this java doc and add the java doc to the ```consumedSubpartitionIndex``` filed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 6]
```diff
@@ -1073,7 +1085,32 @@ private boolean queueChannelUnsafe(InputChannel channel, boolean priority) {
...
+            this.subpartitionIndex = subpartitionIndex;
+        }
+
+        @Override
+        public int hashCode() {
+            return partitionID.hashCode() ^ subpartitionIndex;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:1107]: nit: maybe return true directly if this == obj.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java [hunk 7]
```diff
@@ -224,15 +245,25 @@ private InputChannel createInputChannel(
...
                                 index,
                                 nettyShuffleDescriptor,
+                                consumedSubpartitionIndex,
                                 channelStatistics,
                                 metrics));
     }
 
+    private static int calculateNumChannels(
+            int numShuffleDescriptors, SubpartitionIndexRange subpartitionIndexRange) {
+        return numShuffleDescriptors
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:255]: Do we need a overflow check?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 8]
```diff
@@ -1073,7 +1085,32 @@ private boolean queueChannelUnsafe(InputChannel channel, boolean priority) {
...
+    public Map<SubpartitionInfo, InputChannel> getInputChannels() {
         return inputChannels;
     }
+
+    static class SubpartitionInfo {
+        private final IntermediateResultPartitionID partitionID;
+        private final int subpartitionIndex;
+
+        SubpartitionInfo(IntermediateResultPartitionID partitionID, int subpartitionIndex) {
+            this.partitionID = partitionID;
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:1097]: nit: maybe add a not null check

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 9]
```diff
@@ -1073,7 +1085,32 @@ private boolean queueChannelUnsafe(InputChannel channel, boolean priority) {
...
+            this.subpartitionIndex = subpartitionIndex;
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:1098]: nit: maybe add an >=0 argument check

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 10]
```diff
@@ -630,9 +601,9 @@ public void testRequestBackoffConfiguration() throws Exception {
             closer.register(netEnv::close);
             closer.register(gate::close);
 
-            assertEquals(gateDesc.getConsumedPartitionType(), gate.getConsumedPartitionType());
+            assertEquals(ResultPartitionType.PIPELINED, gate.getConsumedPartitionType());
```
Comment (wsry) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:604]: Is it a bugfix?

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 11]
```diff
@@ -834,14 +817,82 @@ public void testUpdateUnknownInputChannel() throws Exception {
...
-                    inputGate.getInputChannels().get(localResultPartitionId.getPartitionId()),
+                    inputGate
+                            .getInputChannels()
+                            .get(createSubpartitionInfo(localResultPartitionId.getPartitionId())),
                     is(instanceOf((LocalInputChannel.class))));
         }
     }
 
+    @Test
+    public void testUpdateUnknownChannelWithSubpartitionIndexRange()
```
Comment (wsry) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:833]: Maybe it's better to also cover the partition request with subpartition range scenario.

**Comment by wsry** [comment]
@wanglijie95 Thanks for the contribution. The code looks generally good to me and I only left several tiny comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java (hunk 4 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:104]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java (hunk 5 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java (hunk 6 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:1108]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java (hunk 7 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java]: Added.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java (hunk 8 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java (hunk 9 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:1099]: Done.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java (hunk 11 above)
Comment (wanglijie95) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java]: Done

**Comment by wanglijie95** [comment]
> @wanglijie95 Thanks for the contribution. The code looks generally good to me and I only left several tiny comments.

Thanks for review @wsry  . I've addressed all your comments. Please help to review again.

**Comment by wanglijie95** [comment]
Squashed commits.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Comment by wsry** [comment]
@wanglijie95 Thanks for the update. Changes LGTM.

**Comment by zhuzhurk** [comment]
Merging.

**Pull request closed without merging**